packages/roosterjs-content-model-api/lib/modelApi/common/queryContentModel.ts

```diff
@@ -0,0 +1,116 @@
+import type {
+    ContentModelBlockType,
+    ContentModelSegmentType,
+    ReadonlyContentModelBlock,
+    ReadonlyContentModelBlockGroup,
+    ReadonlyContentModelParagraph,
+    ReadonlyContentModelSegment,
+    ReadonlyContentModelTable,
+} from 'roosterjs-content-model-types';
+
+/**
+ * Options for queryContentModel
+ */
+export interface QueryContentModelOptions<T> {
+    /**
+     * The type of block to query @default 'Paragraph'
+     */
+    type?: ContentModelBlockType;
+
+    /**
+     * The type of segment to query
+     */
+    segmentType?: ContentModelSegmentType;
+
+    /**
+     * Optional selector to filter the blocks/segments
+     */
+    selector?: (element: T) => boolean;
+
+    /**
+     * True to return the first block only, false to return all blocks
+     */
+    findFirstOnly?: boolean;
+}
+
+/**
+ * Query content model blocks or segments
+ * @param group The block group to query
+ * @param options The query option
+ */
+export function queryContentModel<
+    T extends ReadonlyContentModelBlock | ReadonlyContentModelSegment
+>(group: ReadonlyContentModelBlockGroup, options: QueryContentModelOptions<T>): T[] {
```

Comment:
I think it will be helpful if we can also return the path to the found element, since in CM we don't have a property to access the parent element. That's how we implement it on OM's repo.

Reply:
That seems a good idea, do you have an example of the implementation?

Reply:
This is basically what we're using now
```typescript
export function findElement<Target extends ReadonlyContentModelType>(
  element: ReadonlyContentModelType,
  path: (ReadonlyContentModelBlockGroup | ReadonlyContentModelParagraph)[],
  match: (element: ReadonlyContentModelType) => element is Target,
  matchAll: boolean,
): ModelElementAndPath<Target>[] {
  const result: ModelElementAndPath<Target>[] = [];
  internalFindElement(element, path, match, matchAll, result);
  return result;
}

function internalFindElement<Target extends ReadonlyContentModelType>(
  element: ReadonlyContentModelType,
  path: (ReadonlyContentModelBlockGroup | ReadonlyContentModelParagraph)[],
  match: (element: ReadonlyContentModelType) => element is Target,
  matchAll: boolean,
  result: ModelElementAndPath<Target>[],
): void {
  if (!matchAll && result.length > 0) {
    return;
  }

  if (match(element)) {
    result.push({ element, path });
    if (!matchAll) {
      return;
    }
  }

  if (isContentModelBlockGroup(element)) {
    element.blocks.forEach((child) => {
      internalFindElement(child, [element, ...path], match, matchAll, result);
    });
    return;
  }

  if (isContentModelBlock(element)) {
    switch (element.blockType) {
      case 'Paragraph':
        element.segments.forEach((segment) => {
          internalFindElement(segment, [element, ...path], match, matchAll, result);
        });
        break;
      case 'Table':
        element.rows.forEach((row) => {
          row.cells.forEach((cell) => {
            internalFindElement(cell, path, match, matchAll, result);
          });
        });
        break;
      default:
        break;
    }
  }
}
```

Reply:
I think that is a nice idea, since sometimes we need the parent block, but I don't want to add more complexity for this function, and I don't want to create a specific return type just this function. For now, I think we can use the filter callback to get the parent and the segment as I did in the findImageEditing function. Also, the user can always create a more customized function to query the need content model blocks/segments. @JiuqingSong any thoughts?